### ISSUE TITLE
Explain diff gutter symbols

### ISSUE THREAD

**Issue by wmertens** [avajs/ava#1558]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2460%20Funded-%2300A156.svg" />](https://issuehunt.io/r/avajs/ava/issues/1558)

Please see this comment: https://github.com/avajs/ava/issues/1558#issuecomment-469031454

Original issue follows.

---

## Description

When using `t.snapshot(foo)` and `foo` does not match the snapshot, the diff that is shown uses `-` to indicate the new `foo` values and `+` to show the old `foo` values.

This is unintuitive.

I believe the problem may be at https://github.com/avajs/ava/blob/14f7095d25abc5ffbff7efd7db962eaf5e86daab/lib/assert.js#L321

### Test Source

```js
import test from 'ava'
test('snapshot diff', t => {
  t.snapshot(`\n${Date.now()}\n`, 'the - value should be less than the + value')
})
```

## Environment

Latest stable ava on OS X

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#2199 Added gutter descriptors](https://issuehunt.io/r/avajs/ava/pull/2199)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by novemberborn** [comment]
The idea is that the new value should be adjusted to match the old value. If you'd used `t.is()`:

```
  direct diff

  /private/var/folders/2_/qczp184x76b2nl034sq5hvxw0000gn/T/tmp.dcvW5BUyA5/test.js:8

   7: test('direct diff', t => {
   8:   t.is(`\n${Date.now() - 42e5}\n`, `\n${Date.now()}\n`)
   9: })

  Difference:

    `␊
  - 1508765718717␊
  + 1508769918717␊
    `
```

What is definitely wrong though is that the colors are not inverted the way the gutters are. I've just opened https://github.com/concordancejs/concordance/issues/40 for that.

If that were fixed would you be happy with the current behavior?

**Comment by wmertens** [comment]
Well, the colors are the primary indicator for me, but I am so used to the
patch view that reversing the order is odd to me, `-` means "previous", not
"wrong value that you should fix".

**Comment by novemberborn** [comment]
In the `t.is(actual, expected)` example below, do the indicators make sense?

![screen shot 2017-10-23 at 15 48 57](https://user-images.githubusercontent.com/33538/31895668-b6c88064-b809-11e7-9052-883a96029efc.png)

The `-` value is supposed to be "bad", with the `+` value being "good." That should then translate to the snapshot, where the current value is assumed to be "bad", since it doesn't match the "good" snapshot.

Just like a code patch in a bugfix takes you from a "bad" file to a "good" file, so would "applying" the diff in AVA's output help you make the actual value match the expectation. I think that holds for snapshots, too.

**Comment by nesbocaj** [comment]
I have to agree with @wmertens, this is counter intuitive.

Reason being that Ava has this nifty feature `$ ava --update-snapshots` which at least in my mind is analogous to `git commit` or `git push`.
With this in mind I find it more intuitive that the diff shows what will happen if the snapshot is updated.
Which means the original should be denoted by a minus and in red, while the new (updated) value should be denoted by a plus and in green.

But that's just my opinion.

**Comment by novemberborn** [comment]
@nesbocaj so if I'm interpreting this right, we'd keep the `-` and `+` as they are now but make the `-` lines green and the `+` lines red?

**Comment by nesbocaj** [comment]
@novemberborn I'm having a hard time figuring out what the current order of the values are, so just to avoid any confusion I'm gonna lay it out below:

\- the snapshot value, colored in green
\+ the new (updated) value, colored in red

Doing it this way will make Ava more approachable to developers coming from Jest and vice versa, as well as work towards establishing some kind of standard in the matter.

**Comment by nesbocaj** [comment]
It is worth noting however that this way of doing it is still different from what most developers will be familiar with from eg. git diffs:

```diff
- the snapshot value
+ the new (updated) value
```

So I think it would be beneficial to have some kind of introduction to the diff syntax, akin to how Jest does it:

<img width="1042" alt="failedsnapshottest" src="https://user-images.githubusercontent.com/2920084/33319652-bdbd68b0-d43f-11e7-97c6-fec73846a941.png">

Of course the lines

```
Received value does not match stored snapshot 1.

- Snapshot
- Received
```

do take up quite a bit of screen real estate and it's up to the Ava developers as to whether or not they want to make this compromise.
Maybe it should only be shown to new developers, eg. the first 5 or so times a snapshot failure is encountered.

**Comment by rzec-r7** [comment]
@novemberborn any update on this? I have been tripped up by the `+` / `-` difference with AVA so many times.

I think what @nesbocaj said makes sense. For me the colors are not as important as the `-` / `+` and since snapshots are tests that are just diffing 2 values, it make the most sense to have the `-` be the existing value in the snapshot and the `+` be the new value from the test (I have not opinion on the colors, probably makes to to just match what jest does).

**Comment by wmertens** [comment]
Yes, the way Jest does it looks good!

**Comment by vadimdemedes** [comment]
Does Jest "explain" the - and + in each diff or once per output?

**Comment by arteniioleg** [comment]
Git and markdown \```diff ...``` shows new changes with green + and deleted with red -
Jest inverted diff colors is confusing.

**Comment by nesbocaj** [comment]
@arteniioleg Jest should only be confusing if you assume the colors have the same purpose, they don't.
In Git the colors signifies changes, with green signifying added/new, and red, removed/old. Whereas in Jest the colors signify correctness of the value, with red signifying correct, and red, incorrect.

In both instances the preceding symbol signifies changes, with `+` signifying added/new, and `-`, removed/old.

It's also worth noting that Jest goes quite far in explaining (or at least trying to explain) this concept to the user.

**Comment by maxrimue** [comment]
Not 100% relating, but I just spent so much time looking into my Unit Tests because I didn't realize what each symbol stood for in this output for a `deepEqual`:
```
Difference:

    [
  -   NaN,
  -   NaN,
  +   12,
  +   30,
    ]
```

Why don't we add a simple
```
  - received
  + expected
```
below? Only the symbols are not completely intuitive to me.

**Comment by novemberborn** [comment]
Let's start with adding `- Received\n+ Expected` when we show diffs, and if they're from a snapshot show `- Received\n+ Snapshot`. We should match the current colors, so the "received" line should be red and the expected/snapshot line green.

In essence, this line should track whether the diff comes from `t.snapshot()`:

https://github.com/avajs/ava/blame/334e15b4af06492c9aed2800a0764f245d6a908b/lib/assert.js#L12

Then the message can be added here:

https://github.com/avajs/ava/blame/334e15b4af06492c9aed2800a0764f245d6a908b/lib/reporters/format-serialized-error.js#L16

**Comment by IssueHuntBot** [comment]
@issuehunt has funded $60.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/26820798/issues/1558) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/26820798/issues/1558).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by kire73** [comment]
I submitted a pull request: #2199 

I just modified the concordance opts in the file:
```
lib\concordance-options.js
```
![image](https://user-images.githubusercontent.com/17581155/62329590-3d08d480-b46b-11e9-9d3a-586af18dc99c.png)

Changes the gutters to:
![image](https://user-images.githubusercontent.com/17581155/62329496-e8fdf000-b46a-11e9-96e9-454883d17e88.png)


Except I added another commit after the fact to correct my atrocious spelling of the word received.

If this is satisfactory enough, I'd forego that bounty, but I'm not exactly sure how to do that. Just looking to make contributions.
